### PR TITLE
GORA-565: Enable Spark in Unit Tests

### PR DESCRIPTION
**Pull request by sneceesay77**
Fixed Spark Unit test, the main issue was the Spark word count program was not properly written. A flat map functionality was missing.

### CODE REVIEW

File: gora-core/src/examples/java/org/apache/gora/examples/WebPageDataCreator.java [hunk 1]
```diff
@@ -64,7 +64,7 @@
   }
   
   public static final String[] CONTENTS = {
-    null,
+    //null,
```
Comment (djkevincr) [inline gora-core/src/examples/java/org/apache/gora/examples/WebPageDataCreator.java]: Why are we ignoring the content = = null case? ( We should see whether any test fail for this case as well. )

File: gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java [hunk 2]
```diff
@@ -31,122 +34,151 @@
...
+			log.info(key);
+			log.info(tokenDatumMap.get(key).toString());
+		}
+		log.info("SparkWordCount debug purpose TokenDatum print ends:");
+
+		// write output to datastore
+		System.out.println(reducedGoraRdd.collect());
+		Configuration sparkHadoopConf = goraSparkEngine.generateOutputConf(outStore);
+		reducedGoraRdd.saveAsNewAPIHadoopDataset(sparkHadoopConf);
+		System.out.println(outStore.get("d"));
```
Comment (djkevincr) [inline gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java]: Shall we remove these, test println()?

File: gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java [hunk 3]
```diff
@@ -31,122 +34,151 @@
...
+		return 1;
+	}
+
+	public int run(String[] args) throws Exception {
+
```
Comment (djkevincr) [inline gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java]: Please reformat this code for 2 space indentation.

File: gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java [hunk 4]
```diff
@@ -31,122 +34,151 @@
...
+		c[0] = inStore.getPersistentClass();
+		sparkConf.registerKryoClasses(c);
+		//
+		JavaSparkContext sc = new JavaSparkContext(sparkConf);
+
+		JavaPairRDD<String, WebPage> goraRDD = goraSparkEngine.initialize(sc, inStore);
+
+		JavaPairRDD<String, String> goraRDDFlatenned = goraRDD.flatMapValues(flatMapFun);
+
+		// Map<String, WebPage> debug = goraRDD.collectAsMap();
```
Comment (djkevincr) [inline gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java]: This commented code can be removed?

File: gora-core/src/examples/java/org/apache/gora/examples/WebPageDataCreator.java (hunk 1 above)
Comment (sneceesay77) [reply gora-core/src/examples/java/org/apache/gora/examples/WebPageDataCreator.java]: this is now fixed.

File: gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java (hunk 2 above)
Comment (sneceesay77) [reply gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java]: all println() debug messages removed.

File: gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java (hunk 3 above)
Comment (sneceesay77) [reply gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java]: Done

File: gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java (hunk 4 above)
Comment (sneceesay77) [reply gora-core/src/examples/java/org/apache/gora/examples/spark/SparkWordCount.java]: Completed.

**Comment by sneceesay77** [comment]
Latest changes pushed.

**Comment by djkevincr** [comment]
@sneceesay77 This looks great, thank you for the contribution. I ll keep this PR for a while, so that others can have a look.

**Comment by djkevincr** [comment]
@sneceesay77 I have locally tested your PR and ran into some issues.
```
# Don't override properties coming from Hadoop configuration for test
# Those properties will contains override for Mongo server port
gora.mongodb.override_hadoop_configuration=true
```
I noticed you have made the following property change and due to this, Apache Flink word count test is failing in mongo-db module. As I can see from the log traces, while running Flink jobs it fails to connect to the Mongo server running on the correct port. Basically we should not override the mongo server port which is set to Hadoop conf ( conf.set(MongoStoreParameters.PROP_MONGO_SERVERS, "127.0.0.1:" + port); ) in GoraMongodbTestDriver from the port added in the property file. ( Value in Hadoop conf should have the priority ) Can you please check on this?

**Comment by djkevincr** [comment]
```
public <K, V extends Persistent> Configuration generateOutputConf(DataStore<K, V> dataStore)
       throws IOException {

      Configuration hadoopConf = ((Configurable) dataStore).getConf(); // Configuration hadoopConf = new Configuration(); -  previously it was
      GoraMapReduceUtils.setIOSerializations(hadoopConf, true);
      Job job = Job.getInstance(hadoopConf);

      return generateOutputConf(job, dataStore.getClass(), dataStore.getKeyClass(),
           dataStore.getPersistentClass());
    }
```
To properly fix the issue, I think GoraSparkEngine class method generateOutputConf should be changed to the above. ( Check java single line comment ) Basically we should reuse the Hadoop conf created at GoraMongodbTestDriver through out entire tests. Initial test startup phase, we set these mongo server details to Hadoop conf created.
I am not sure whether there exist a case which returns dataStore conf as null. ( Eg:- de serilizations ) so we can do something similar to below. This code is extreacted from the GoraSparkEngine class method initialize. Can you please address the changes and update the PR?
```
if ((dataStore instanceof Configurable)
        && ((Configurable) dataStore).getConf() != null) {
      hadoopConf = ((Configurable) dataStore).getConf();
    } else {
      hadoopConf = new Configuration();
    }
```

**Comment by djkevincr** [comment]
I will merge this PR as it is and send a new PR to fix the failing tests.

**Pull request merged**